Added model summary and risk assessment for commands that violate sandbox policy

etraut-openai opened the pull request:
This PR adds support for a model-based summary and risk assessment for commands that violate the sandbox policy and require user approval. This aids the user in evaluating whether the command should be approved.

The feature works by taking a failed command and passing it back to the model and asking it to summarize the command, give it a risk level (low, medium, high) and a risk category (e.g. "data deletion" or "data exfiltration"). It uses a new conversation thread so the context in the existing thread doesn't influence the answer. If the call to the model fails or takes longer than 5 seconds, it falls back to the current behavior.

For now, this is an experimental feature and is gated by a config key `experimental_sandbox_command_assessment`.

Here is a screen shot of the approval prompt showing the risk assessment and summary.

<img width="723" height="282" alt="image" src="https://github.com/user-attachments/assets/4597dd7c-d5a0-4e9f-9d13-414bd082fd6b" />

etraut-openai committed:
Added model summary and risk assessment for commands that violate sandbox policy and require user approval

etraut-openai committed:
Refined TUI

etraut-openai committed:
Removed inaccurate part of prompt

jif-oai commented on codex-rs/core/src/sandboxing/assessment.rs:
```diff
@@ -0,0 +1,318 @@
+
+    let platform = std::env::consts::OS;
+    let mut prompt_sections = Vec::new();
+    prompt_sections.push(format!("Platform: {platform}"));
```
Can we use a file + askama for this. Would be cleaner IMO

pakrym-oai commented on codex-rs/protocol/src/protocol.rs:
```diff
@@ -1126,6 +1126,58 @@ pub struct ExecCommandOutputDeltaEvent {
     pub chunk: Vec<u8>,
 }
 
+#[derive(Debug, Clone, Copy, Deserialize, Serialize, PartialEq, Eq, Hash, JsonSchema, TS)]
```
nit: worth to start splitting out into separate files? `risk.rs` maybe? protocol.rs is getting large.

pakrym-oai commented on codex-rs/core/src/sandboxing/assessment.rs:
```diff
@@ -0,0 +1,318 @@
+    let duration = start.elapsed();
+
+    match assessment_result {
+        Ok(Ok(Some(raw))) => {
```
nit: This section is very verbose with many `Err` branches being similar. Any way we can simplify.

pakrym-oai reviewed (approved):
LGTM after @jif-oai 's comments.

etraut-openai committed:
Code review feedback:
* Moved prompt into its own file and switched it to use askama for templating
* Refactored sandbox_retry_data trait for simplification
* Fixed otel telemetry so assessment conversation doesn't appear as a new task
* Added otel telemetry point for recording latency of assessment
* Removed defensive JSON parsing of assessment response

Removed new experimental config key from public documentation for now. We're going to roll this out internally first to get feedback.

etraut-openai committed:
More code review feedback:
* Simplified config handling by leveraging "features" mechanism
* Moved approvals-related schemas from protocol.rs to simplify

etraut-openai committed:
Bug fixes

etraut-openai committed:
Merge branch 'main' into etraut/command-assessment

etraut-openai commented:
@codex review
